In [6]:
import pandas as pd

## Let's see how to import data from CSV, API(JSON),SQL Database...etc and hold the data in a dataframe object

In [7]:
df_movies =pd.read_csv('Data/IMDB movies.csv')

## Display a maximum of 5 columns

In [8]:
df_movies.head()

imdb_title_id                                              title  \
0     tt0000574                        The Story of the Kelly Gang   
1     tt0001892                                     Den sorte drøm   
2     tt0002101                                          Cleopatra   
3     tt0002130                                          L'Inferno   
4     tt0002199  From the Manger to the Cross; or, Jesus of Naz...   

                                      original_title  year date_published  \
0                        The Story of the Kelly Gang  1906     1906-12-26   
1                                     Den sorte drøm  1911     1911-08-19   
2                                          Cleopatra  1912     1912-11-13   
3                                          L'Inferno  1911     1911-03-06   
4  From the Manger to the Cross; or, Jesus of Naz...  1912           1913   

                       genre  duration           country language  \
0    Biography, Crime, Drama        70         Australia      NaN   
1                      Drama        53  Germany, Denmark      NaN   
2             Drama, History       100               USA  English   
3  Adventure, Drama, Fantasy        68             Italy  Italian   
4           Biography, Drama        60               USA  English   

                              director  ...  \
0                         Charles Tait  ...   
1                            Urban Gad  ...   
2                   Charles L. Gaskill  ...   
3  Francesco Bertolini, Adolfo Padovan  ...   
4                        Sidney Olcott  ...   

                                              actors  \
0  Elizabeth Tait, John Tait, Norman Campbell, Be...   
1  Asta Nielsen, Valdemar Psilander, Gunnar Helse...   
2  Helen Gardner, Pearl Sindelar, Miss Fielding, ...   
3  Salvatore Papa, Arturo Pirovano, Giuseppe de L...   
4  R. Henderson Bland, Percy Dyer, Gene Gauntier,...   

                                         description avg_vote votes   budget  \
0  True story of notorious Australian outlaw Ned ...      6.1   537   $ 2250   
1  Two men of high rank are both wooing the beaut...      5.9   171      NaN   
2  The fabled queen of Egypt's affair with Roman ...      5.2   420  $ 45000   
3  Loosely adapted from Dante's Divine Comedy and...      7.0  2019      NaN   
4  An account of the life of Jesus Christ, based ...      5.7   438      NaN   

   usa_gross_income worlwide_gross_income metascore reviews_from_users  \
0               NaN                   NaN       NaN                7.0   
1               NaN                   NaN       NaN                4.0   
2               NaN                   NaN       NaN               24.0   
3               NaN                   NaN       NaN               28.0   
4               NaN                   NaN       NaN               12.0   

   reviews_from_critics  
0                   7.0  
1                   2.0  
2                   3.0  
3                  14.0  
4                   5.0  

[5 rows x 22 columns]

# #Locate the data in the various rows using .loc function

In [9]:
df_movies.loc[4]

imdb_title_id                                                    tt0002199
title                    From the Manger to the Cross; or, Jesus of Naz...
original_title           From the Manger to the Cross; or, Jesus of Naz...
year                                                                  1912
date_published                                                        1913
genre                                                     Biography, Drama
duration                                                                60
country                                                                USA
language                                                           English
director                                                     Sidney Olcott
writer                                                       Gene Gauntier
production_company                                           Kalem Company
actors                   R. Henderson Bland, Percy Dyer, Gene Gauntier,...
description              

# Import new csv creating an index this time around using column imdb_title_id 
##The Imdb_title_id will no longer be a column in our data

In [10]:
df_movies = pd.read_csv("Data/IMDb movies.csv", index_col = 'imdb_title_id')

In [11]:
df_movies.head(2)

title               original_title  year  \
imdb_title_id                                                                   
tt0000574      The Story of the Kelly Gang  The Story of the Kelly Gang  1906   
tt0001892                   Den sorte drøm               Den sorte drøm  1911   

              date_published                    genre  duration  \
imdb_title_id                                                     
tt0000574         1906-12-26  Biography, Crime, Drama        70   
tt0001892         1911-08-19                    Drama        53   

                        country language      director  \
imdb_title_id                                            
tt0000574             Australia      NaN  Charles Tait   
tt0001892      Germany, Denmark      NaN     Urban Gad   

                                              writer  ...  \
imdb_title_id                                         ...   
tt0000574                               Charles Tait  ...   
tt0001892      Urban Gad, Gebhard Schätzler-Perasini  ...   

                                                          actors  \
imdb_title_id                                                      
tt0000574      Elizabeth Tait, John Tait, Norman Campbell, Be...   
tt0001892      Asta Nielsen, Valdemar Psilander, Gunnar Helse...   

                                                     description avg_vote  \
imdb_title_id                                                               
tt0000574      True story of notorious Australian outlaw Ned ...      6.1   
tt0001892      Two men of high rank are both wooing the beaut...      5.9   

               votes  budget usa_gross_income worlwide_gross_income metascore  \
imdb_title_id                                                                   
tt0000574        537  $ 2250              NaN                   NaN       NaN   
tt0001892        171     NaN              NaN                   NaN       NaN   

               reviews_from_users  reviews_from_critics  
imdb_title_id                                            
tt0000574                     7.0                   7.0  
tt0001892                     4.0                   2.0  

[2 rows x 21 columns]

## Importing Data from a REST API (i.e JSON)
_In python we decode JSON into dictionaries

In [12]:
df_posts = pd.read_json("https://jsonplaceholder.typicode.com/posts")

In [13]:
df_posts ['id'].count() == len(set(df_posts['id']))

True

In [14]:
df_posts.head(1)

userId  id                                              title  \
0       1   1  sunt aut facere repellat provident occaecati e...   

                                                body  
0  quia et suscipit\nsuscipit recusandae consequu...

In [15]:
df_posts.set_index('id', inplace=True)

In [16]:
df_posts.head()

userId                                              title  \
id                                                              
1        1  sunt aut facere repellat provident occaecati e...   
2        1                                       qui est esse   
3        1  ea molestias quasi exercitationem repellat qui...   
4        1                               eum et est occaecati   
5        1                                 nesciunt quas odio   

                                                 body  
id                                                     
1   quia et suscipit\nsuscipit recusandae consequu...  
2   est rerum tempore vitae\nsequi sint nihil repr...  
3   et iusto sed quo iure\nvoluptatem occaecati om...  
4   ullam et saepe reiciendis voluptatem adipisci\...  
5   repudiandae veniam quaerat sunt sed\nalias aut...

In [17]:
!pip install psycopg2

## Exploratory Data Analysis i.e Descriptive Statistics . Data preparation/cleaning Steps are:
*Run df.info() to check column/row frequencies of null/non null values.
*If more than half of the rows in a column has None/NA/Null Values then we could drop the column.
*If the null values are few, then we can replace them with mean/median/mode
*Deal with duplicate values.Remove and keep only one.
## Performing Descriptive Statistics
*Also run describe() method on your df.This will show frequency/mean/min/max/standard deviation/variance of continuous values(numbers only)
*Check the relationship between continous variables i.e. one column and another

## Data visualization (Charts)

In [18]:
df_movies.info()

<class 'pandas.core.frame.DataFrame'>
Index: 81273 entries, tt0000574 to tt9914286
Data columns (total 21 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   title                  81273 non-null  object 
 1   original_title         81273 non-null  object 
 2   year                   81273 non-null  int64  
 3   date_published         81273 non-null  object 
 4   genre                  81273 non-null  object 
 5   duration               81273 non-null  int64  
 6   country                81234 non-null  object 
 7   language               80518 non-null  object 
 8   director               81200 non-null  object 
 9   writer                 79780 non-null  object 
 10  production_company     76948 non-null  object 
 11  actors                 81207 non-null  object 
 12  description            78843 non-null  object 
 13  avg_vote               81273 non-null  float64
 14  votes                  81273 non-null  int64  


# How to replace the null values in Country

##We replace the values with a measure of tendency since they are few . In this case we choose mode(most common country)because the data type is object(string)

_Budget column has 22804 out of 81273 values which are null. How would we deal with this scenario?

We would drop the column which contains too many null values

Reviews_from_users_ We would replace the null values with the mean if The data type is an integer.

In [19]:
df_movies.columns

Index(['title', 'original_title', 'year', 'date_published', 'genre',
       'duration', 'country', 'language', 'director', 'writer',
       'production_company', 'actors', 'description', 'avg_vote', 'votes',
       'budget', 'usa_gross_income', 'worlwide_gross_income', 'metascore',
       'reviews_from_users', 'reviews_from_critics'],
      dtype='object')

#### 1.Clean up our columns and check for duplicates

In [20]:
df_movies.head(1)

title               original_title  year  \
imdb_title_id                                                                   
tt0000574      The Story of the Kelly Gang  The Story of the Kelly Gang  1906   

              date_published                    genre  duration    country  \
imdb_title_id                                                                
tt0000574         1906-12-26  Biography, Crime, Drama        70  Australia   

              language      director        writer  ...  \
imdb_title_id                                       ...   
tt0000574          NaN  Charles Tait  Charles Tait  ...   

                                                          actors  \
imdb_title_id                                                      
tt0000574      Elizabeth Tait, John Tait, Norman Campbell, Be...   

                                                     description avg_vote  \
imdb_title_id                                                               
tt0000574      True story of notorious Australian outlaw Ned ...      6.1   

               votes  budget usa_gross_income worlwide_gross_income metascore  \
imdb_title_id                                                                   
tt0000574        537  $ 2250              NaN                   NaN       NaN   

               reviews_from_users  reviews_from_critics  
imdb_title_id                                            
tt0000574                     7.0                   7.0  

[1 rows x 21 columns]

In [21]:
df_movies_one = pd.read_csv('Data/imdb_data_1000.csv',index_col='Rank')

In [22]:
df_movies_one.head(1)

Title                    Genre  \
Rank                                                     
1     Guardians of the Galaxy  Action,Adventure,Sci-Fi   

                                            Description    Director  \
Rank                                                                  
1     A group of intergalactic criminals are forced ...  James Gunn   

                                                 Actors  Year  \
Rank                                                            
1     Chris Pratt, Vin Diesel, Bradley Cooper, Zoe S...  2014   

      Runtime (Minutes)  Rating   Votes  Revenue (Millions)  Metascore  
Rank                                                                    
1                   121     8.1  757074              333.13       76.0

In [23]:
df_movies_one.columns

Index(['Title', 'Genre', 'Description', 'Director', 'Actors', 'Year',
       'Runtime (Minutes)', 'Rating', 'Votes', 'Revenue (Millions)',
       'Metascore'],
      dtype='object')

In [24]:

df_movies_one.rename(columns = {'Runtime (Minutes)': 'Run_Minutes',
                                'Revenue (Millions)': 'Revenue_Millions'}, inplace = True)

In [25]:
df_movies_one.columns = [c.lower()for c in df_movies_one]

In [26]:
df_movies_one.columns

Index(['title', 'genre', 'description', 'director', 'actors', 'year',
       'run_minutes', 'rating', 'votes', 'revenue_millions', 'metascore'],
      dtype='object')

In [27]:
df_movies_one.shape

(1000, 11)

In [28]:
df_movies_one.drop_duplicates(keep='first', inplace = True)

In [29]:
df_movies_one.shape

(1000, 11)

# Create a temporary data frame, create duplicate values using append method

In [30]:
tmp_df =df_movies_one.append(df_movies_one)

In [31]:
tmp_df = tmp_df.append(df_movies_one)

In [32]:
tmp_df.shape

(3000, 11)

In [33]:
tmp_df_one = tmp_df.copy()

In [34]:
tmp_df.drop_duplicates(inplace = True,keep ='last')

In [35]:
tmp_df.shape

(1000, 11)

In [36]:
tmp_df_one.drop_duplicates(inplace = True, keep =False)

In [37]:
tmp_df_one.shape

(0, 11)

In [38]:
df_movies.shape

(81273, 21)

In [39]:
df_movies_one.shape

(1000, 11)

In [40]:
df_comb = df_movies_one.append(df_movies)

In [41]:
df_comb.shape

(82273, 24)

In [42]:
df_comb.columns

Index(['title', 'genre', 'description', 'director', 'actors', 'year',
       'run_minutes', 'rating', 'votes', 'revenue_millions', 'metascore',
       'original_title', 'date_published', 'duration', 'country', 'language',
       'writer', 'production_company', 'avg_vote', 'budget',
       'usa_gross_income', 'worlwide_gross_income', 'reviews_from_users',
       'reviews_from_critics'],
      dtype='object')

In [43]:
df_movies_one.shape

(1000, 11)

In [44]:
df_movies.shape

(81273, 21)

In [45]:
df_nulls = df_movies_one.isnull().sum()

In [46]:
df_nulls

title                 0
genre                 0
description           0
director              0
actors                0
year                  0
run_minutes           0
rating                0
votes                 0
revenue_millions    128
metascore            64
dtype: int64

In [47]:
df_movies_one1=df_movies_one.dropna()

In [48]:
df_movies_one1.shape

(838, 11)

In [49]:
df_movies_one2 = df_movies_one.copy()

In [50]:
df_movies_one3 = df_movies_one.copy()

## Remove a column with at least 1 null values

In [51]:
df_movies_one2.dropna(axis=1, inplace = True)

In [52]:
df_movies_one2.shape

(1000, 9)

In [53]:
df_movies_one3.dropna(axis=0, inplace = True)

In [54]:
df_movies_one3.shape

(838, 11)

In [55]:
revenue = df_movies_one['revenue_millions']

In [56]:
revenue.head()

Rank
1    333.13
2    126.46
3    138.12
4    270.32
5    325.02
Name: revenue_millions, dtype: float64

In [57]:
type(revenue)

pandas.core.series.Series

In [58]:
d4= revenue.mean()

In [59]:
d4

82.95637614678897

In [60]:
d_null = revenue.isna()

In [61]:
d_null

Rank
1       False
2       False
3       False
4       False
5       False
        ...  
996      True
997     False
998     False
999      True
1000    False
Name: revenue_millions, Length: 1000, dtype: bool

In [62]:
revenue.fillna(d4, inplace = True)

In [63]:
revenue.shape

(1000,)

# Find how many values in series  have the value 'd4' as above

In [64]:
df_movies_one.isnull().sum()

title                0
genre                0
description          0
director             0
actors               0
year                 0
run_minutes          0
rating               0
votes                0
revenue_millions     0
metascore           64
dtype: int64

In [65]:
d5 = df_movies_one.copy()

In [66]:
d5['revenue_millions'].fillna(d5['revenue_millions'].mean(),inplace =True)

In [67]:
d5['metascore'].fillna(d5['metascore'].mean(),inplace =True)

In [68]:
d5.to_csv('data/imdb_1000_clean.csv')

In [69]:
d5.isnull().sum()

title               0
genre               0
description         0
director            0
actors              0
year                0
run_minutes         0
rating              0
votes               0
revenue_millions    0
metascore           0
dtype: int64

In [70]:
d5.isnull()

title  genre  description  director  actors   year  run_minutes  rating  \
Rank                                                                            
1     False  False        False     False   False  False        False   False   
2     False  False        False     False   False  False        False   False   
3     False  False        False     False   False  False        False   False   
4     False  False        False     False   False  False        False   False   
5     False  False        False     False   False  False        False   False   
...     ...    ...          ...       ...     ...    ...          ...     ...   
996   False  False        False     False   False  False        False   False   
997   False  False        False     False   False  False        False   False   
998   False  False        False     False   False  False        False   False   
999   False  False        False     False   False  False        False   False   
1000  False  False        False     False   False  False        False   False   

      votes  revenue_millions  metascore  
Rank                                      
1     False             False      False  
2     False             False      False  
3     False             False      False  
4     False             False      False  
5     False             False      False  
...     ...               ...        ...  
996   False             False      False  
997   False             False      False  
998   False             False      False  
999   False             False      False  
1000  False             False      False  

[1000 rows x 11 columns]